File: sonet-client/ios/Sonet/NativeApp/Networking/SonetGRPCClient.swift
import Foundation
import GRPC
import NIO
import NIOHTTP1
import NIOTransportServices

// MARK: - Sonet gRPC Client
class SonetGRPCClient: ObservableObject {
    // MARK: - Published Properties
    @Published var isConnected = false
    @Published var connectionError: String?
    
    // MARK: - Private Properties
    private var channel: GRPCChannel?
    private let group = PlatformSupport.makeEventLoopGroup(loopCount: 1)
    private let configuration: SonetConfiguration
    
    // MARK: - Service Clients
    private var userServiceClient: UserServiceClient?
    private var noteServiceClient: NoteServiceClient?
    private var timelineServiceClient: TimelineServiceClient?
    private var messagingServiceClient: MessagingServiceClient?
    private var searchServiceClient: SearchServiceClient?
    
    // MARK: - Initialization
    init(configuration: SonetConfiguration) {
        self.configuration = configuration
        setupConnection()
    }
    
    deinit {
        try? channel?.close().wait()
        try? group.syncShutdownGracefully()
    }
    
    // MARK: - Connection Setup
    private func setupConnection() {
        do {
            let channel = try GRPCChannelPool.with(
                target: .hostAndPort(configuration.host, configuration.port),
                transportSecurity: configuration.useTLS ? .tls(GRPCTLSConfiguration.makeClientConfigurationBackedByNIOSSL()) : .plaintext,
                eventLoopGroup: group
            )
            
            self.channel = channel
            setupServiceClients()
            testConnection()
            
        } catch {
            connectionError = "Failed to create gRPC channel: \(error.localizedDescription)"
            print("gRPC Connection Error: \(error)")
        }
    }
    
    private func setupServiceClients() {
        guard let channel = channel else { return }
        
        userServiceClient = UserServiceClient(channel: channel)
        noteServiceClient = NoteServiceClient(channel: channel)
        timelineServiceClient = TimelineServiceClient(channel: channel)
        messagingServiceClient = MessagingServiceClient(channel: channel)
        searchServiceClient = SearchServiceClient(channel: channel)
    }
    
    private func testConnection() {
        // Test connection with a simple call
        Task {
            do {
                let response = try await userServiceClient?.ping(.init())
                isConnected = true
                connectionError = nil
                print("gRPC connection established successfully")
            } catch {
                isConnected = false
                connectionError = "Connection test failed: \(error.localizedDescription)"
                print("gRPC connection test failed: \(error)")
            }
        }
    }
    
    // MARK: - Authentication Methods
    func authenticate(email: String, password: String) async throws -> UserProfile {
        guard let client = userServiceClient else {
            throw SonetError.serviceUnavailable
        }
        
        let request = AuthRequest.with {
            $0.email = email
            $0.password = password
            $0.sessionType = .mobile
        }
        
        let response = try await client.authenticate(request)
        return response.userProfile
    }
    
    func registerUser(username: String, email: String, password: String) async throws -> UserProfile {
        guard let client = userServiceClient else {
            throw SonetError.serviceUnavailable
        }
        
        let request = RegisterUserRequest.with {
            $0.username = username
            $0.email = email
            $0.password = password
        }
        
        let response = try await client.registerUser(request)
        return response.userProfile
    }
    
    func refreshSession(sessionId: String) async throws -> Session {
        guard let client = userServiceClient else {
            throw SonetError.serviceUnavailable
        }
        
        let request = RefreshSessionRequest.with {
            $0.sessionID = sessionId
        }
        
        let response = try await client.refreshSession(request)
        return response.session
    }
    
    // MARK: - Note Methods
    func getHomeTimeline(page: Int, pageSize: Int) async throws -> [TimelineItem] {
        guard let client = timelineServiceClient else {
            throw SonetError.serviceUnavailable
        }
        
        let request = GetTimelineRequest.with {
            $0.algorithm = .hybrid
            $0.page = UInt32(page)
            $0.pageSize = UInt32(pageSize)
        }
        
        let response = try await client.getHomeTimeline(request)
        return response.items
    }
    
    func getNotesBatch(noteIds: [String]) async throws -> [Note] {
        guard let client = noteServiceClient else {
            throw SonetError.serviceUnavailable
        }
        
        let request = GetNotesBatchRequest.with {
            $0.noteIds = noteIds
        }
        
        let response = try await client.getNotesBatch(request)
        return response.notes
    }
    
    func likeNote(noteId: String, userId: String) async throws -> LikeNoteResponse {
        guard let client = noteServiceClient else {
            throw SonetError.serviceUnavailable
        }
        
        let request = LikeNoteRequest.with {
            $0.noteID = noteId
            $0.userID = userId
        }
        
        return try await client.likeNote(request)
    }
    
    func unlikeNote(noteId: String, userId: String) async throws -> UnlikeNoteResponse {
        guard let client = noteServiceClient else {
            throw SonetError.serviceUnavailable
        }
        
        let request = UnlikeNoteRequest.with {
            $0.noteID = noteId
            $0.userID = userId
        }
        
        return try await client.unlikeNote(request)
    }
    
    // MARK: - Search Methods
    func searchUsers(query: String, page: Int, pageSize: Int) async throws -> [UserProfile] {
        guard let client = searchServiceClient else {
            throw SonetError.serviceUnavailable
        }
        
        let request = SearchUsersRequest.with {
            $0.query = query
            $0.page = UInt32(page)
            $0.pageSize = UInt32(pageSize)
        }
        
        let response = try await client.searchUsers(request)
        return response.users
    }
    
    func searchNotes(query: String, page: Int, pageSize: Int) async throws -> [Note] {
        guard let client = searchServiceClient else {
            throw SonetError.serviceUnavailable
        }
        
        let request = SearchNotesRequest.with {
            $0.query = query
            $0.page = UInt32(page)
            $0.pageSize = UInt32(pageSize)
        }
        
        let response = try await client.searchNotes(request)
        return response.notes
    }
    
    // MARK: - Messaging Methods
    func getConversations(userId: String) async throws -> [Conversation] {
        guard let client = messagingServiceClient else {
            throw SonetError.serviceUnavailable
        }
        
        let request = GetConversationsRequest.with {
            $0.userID = userId
        }
        
        let response = try await client.getConversations(request)
        return response.conversations
    }
    
    func getMessages(conversationId: String, page: Int, pageSize: Int) async throws -> [Message] {
        guard let client = messagingServiceClient else {
            throw SonetError.serviceUnavailable
        }
        
        let request = GetMessagesRequest.with {
            $0.conversationID = conversationId
            $0.page = UInt32(page)
            $0.pageSize = UInt32(pageSize)
        }
        
        let response = try await client.getMessages(request)
        return response.messages
    }
}

// MARK: - Configuration
struct SonetConfiguration {
    let host: String
    let port: Int
    let useTLS: Bool
    let timeout: TimeInterval
    
    static let development = SonetConfiguration(
        host: "localhost",
        port: 50051,
        useTLS: false,
        timeout: 30.0
    )
    
    static let staging = SonetConfiguration(
        host: "staging.sonet.app",
        port: 443,
        useTLS: true,
        timeout: 30.0
    )
    
    static let production = SonetConfiguration(
        host: "api.sonet.app",
        port: 443,
        useTLS: true,
        timeout: 30.0
    )
}

// MARK: - Error Types
enum SonetError: LocalizedError {
    case serviceUnavailable
    case authenticationFailed
    case networkError
    case invalidResponse
    case rateLimited
    
    var errorDescription: String? {
        switch self {
        case .serviceUnavailable:
            return "Service is currently unavailable"
        case .authenticationFailed:
            return "Authentication failed"
        case .networkError:
            return "Network connection error"
        case .invalidResponse:
            return "Invalid response from server"
        case .rateLimited:
            return "Too many requests, please try again later"
        }
    }
}

// MARK: - Protocol Buffer Extensions
extension UserProfile {
    var displayName: String {
        return self.displayName.isEmpty ? self.username : self.displayName
    }
    
    var isVerified: Bool {
        return self.isVerified
    }
    
    var avatarURL: URL? {
        return URL(string: self.avatarURL)
    }
}

extension Note {
    var text: String {
        return self.content
    }
    
    var author: UserProfile {
        return self.author
    }
    
    var createdAt: Date {
        return Date(timeIntervalSince1970: TimeInterval(self.createdAt.seconds))
    }
}

File: sonet-client/android/app/src/main/java/xyz/sonet/app/grpc/SonetGRPCClient.kt
package xyz.sonet.app.grpc

import android.content.Context
import io.grpc.ManagedChannel
import io.grpc.ManagedChannelBuilder
import io.grpc.stub.StreamObserver
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.flow
import kotlinx.coroutines.suspendCancellableCoroutine
import kotlin.coroutines.resume
import kotlin.coroutines.resumeWithException
import xyz.sonet.app.grpc.proto.*
import xyz.sonet.app.grpc.proto.UserServiceGrpc
import xyz.sonet.app.grpc.proto.NoteServiceGrpc
import xyz.sonet.app.grpc.proto.TimelineServiceGrpc
import xyz.sonet.app.grpc.proto.SearchServiceGrpc
import xyz.sonet.app.grpc.proto.MessagingServiceGrpc
import java.util.concurrent.TimeUnit

class SonetGRPCClient(
    private val context: Context,
    private val configuration: SonetConfiguration
) {
    
    // MARK: - Properties
    private var channel: ManagedChannel? = null
    private var userServiceClient: UserServiceGrpc.UserServiceBlockingStub? = null
    private var noteServiceClient: NoteServiceGrpc.NoteServiceBlockingStub? = null
    private var timelineServiceClient: TimelineServiceGrpc.TimelineServiceBlockingStub? = null
    private var searchServiceClient: SearchServiceGrpc.SearchServiceBlockingStub? = null
    private var messagingServiceClient: MessagingServiceGrpc.MessagingServiceBlockingStub? = null
    
    // MARK: - Initialization
    init {
        setupConnection()
    }
    
    // MARK: - Connection Setup
    private fun setupConnection() {
        try {
            val channelBuilder = ManagedChannelBuilder
                .forAddress(configuration.host, configuration.port)
                .keepAliveTime(30, TimeUnit.SECONDS)
                .keepAliveTimeout(5, TimeUnit.SECONDS)
                .maxInboundMessageSize(1024 * 1024) // 1MB
            
            if (configuration.useTLS) {
                channelBuilder.useTransportSecurity()
            } else {
                channelBuilder.usePlaintext()
            }
            
            channel = channelBuilder.build()
            setupServiceClients()
            testConnection()
            
        } catch (error: Exception) {
            println("gRPC Connection Error: $error")
        }
    }
    
    private fun setupServiceClients() {
        channel?.let { channel ->
            userServiceClient = UserServiceGrpc.newBlockingStub(channel)
            noteServiceClient = NoteServiceGrpc.newBlockingStub(channel)
            timelineServiceClient = TimelineServiceGrpc.newBlockingStub(channel)
            searchServiceClient = SearchServiceGrpc.newBlockingStub(channel)
            messagingServiceClient = MessagingServiceGrpc.newBlockingStub(channel)
        }
    }
    
    private fun testConnection() {
        // Test connection with a simple ping
        try {
            val request = PingRequest.newBuilder().build()
            val response = userServiceClient?.ping(request)
            println("gRPC connection established successfully")
        } catch (error: Exception) {
            println("gRPC connection test failed: $error")
        }
    }
    
    // MARK: - Authentication Methods
    suspend fun authenticate(email: String, password: String): UserProfile {
        return suspendCancellableCoroutine { continuation ->
            try {
                val request = AuthRequest.newBuilder()
                    .setEmail(email)
                    .setPassword(password)
                    .setSessionType(SessionType.MOBILE)
                    .build()
                
                val response = userServiceClient?.authenticate(request)
                    ?: throw Exception("User service client not available")
                
                continuation.resume(response.userProfile)
                
            } catch (error: Exception) {
                continuation.resumeWithException(error)
            }
        }
    }
    
    suspend fun registerUser(username: String, email: String, password: String): UserProfile {
        return suspendCancellableCoroutine { continuation ->
            try {
                val request = RegisterUserRequest.newBuilder()
                    .setUsername(username)
                    .setEmail(email)
                    .setPassword(password)
                    .build()
                
                val response = userServiceClient?.registerUser(request)
                    ?: throw Exception("User service client not available")
                
                continuation.resume(response.userProfile)
                
            } catch (error: Exception) {
                continuation.resumeWithException(error)
            }
        }
    }
    
    suspend fun refreshSession(sessionId: String): Session {
        return suspendCancellableCoroutine { continuation ->
            try {
                val request = RefreshSessionRequest.newBuilder()
                    .setSessionID(sessionId)
                    .build()
                
                val response = userServiceClient?.refreshSession(request)
                    ?: throw Exception("User service client not available")
                
                continuation.resume(response.session)
                
            } catch (error: Exception) {
                continuation.resumeWithException(error)
            }
        }
    }
    
    // MARK: - Note Methods
    suspend fun getHomeTimeline(page: Int, pageSize: Int): List<TimelineItem> {
        return suspendCancellableCoroutine { continuation ->
            try {
                val request = GetTimelineRequest.newBuilder()
                    .setAlgorithm(TimelineAlgorithm.HYBRID)
                    .setPage(page.toLong())
                    .setPageSize(pageSize.toLong())
                    .build()
                
                val response = timelineServiceClient?.getHomeTimeline(request)
                    ?: throw Exception("Timeline service client not available")
                
                continuation.resume(response.itemsList)
                
            } catch (error: Exception) {
                continuation.resumeWithException(error)
            }
        }
    }
    
    suspend fun getNotesBatch(noteIds: List<String>): List<Note> {
        return suspendCancellableCoroutine { continuation ->
            try {
                val request = GetNotesBatchRequest.newBuilder()
                    .addAllNoteIds(noteIds)
                    .build()
                
                val response = noteServiceClient?.getNotesBatch(request)
                    ?: throw Exception("Note service client not available")
                
                continuation.resume(response.notesList)
                
            } catch (error: Exception) {
                continuation.resumeWithException(error)
            }
        }
    }
    
    suspend fun likeNote(noteId: String, userId: String): LikeNoteResponse {
        return suspendCancellableCoroutine { continuation ->
            try {
                val request = LikeNoteRequest.newBuilder()
                    .setNoteID(noteId)
                    .setUserID(userId)
                    .build()
                
                val response = noteServiceClient?.likeNote(request)
                    ?: throw Exception("Note service client not available")
                
                continuation.resume(response)
                
            } catch (error: Exception) {
                continuation.resumeWithException(error)
            }
        }
    }
    
    suspend fun unlikeNote(noteId: String, userId: String): UnlikeNoteResponse {
        return suspendCancellableCoroutine { continuation ->
            try {
                val request = UnlikeNoteRequest.newBuilder()
                    .setNoteID(noteId)
                    .setUserID(userId)
                    .build()
                
                val response = noteServiceClient?.unlikeNote(request)
                    ?: throw Exception("Note service client not available")
                
                continuation.resume(response)
                
            } catch (error: Exception) {
                continuation.resumeWithException(error)
            }
        }
    }
    
    // MARK: - Search Methods
    suspend fun searchUsers(query: String, page: Int, pageSize: Int): List<UserProfile> {
        return suspendCancellableCoroutine { continuation ->
            try {
                val request = SearchUsersRequest.newBuilder()
                    .setQuery(query)
                    .setPage(page.toLong())
                    .setPageSize(pageSize.toLong())
                    .build()
                
                val response = searchServiceClient?.searchUsers(request)
                    ?: throw Exception("Search service client not available")
                
                continuation.resume(response.usersList)
                
            } catch (error: Exception) {
                continuation.resumeWithException(error)
            }
        }
    }
    
    suspend fun searchNotes(query: String, page: Int, pageSize: Int): List<Note> {
        return suspendCancellableCoroutine { continuation ->
            try {
                val request = SearchNotesRequest.newBuilder()
                    .setQuery(query)
                    .setPage(page.toLong())
                    .setPageSize(pageSize.toLong())
                    .build()
                
                val response = searchServiceClient?.searchNotes(request)
                    ?: throw Exception("Search service client not available")
                
                continuation.resume(response.notesList)
                
            } catch (error: Exception) {
                continuation.resumeWithException(error)
            }
        }
    }
    
    // MARK: - Messaging Methods
    suspend fun getConversations(userId: String): List<Conversation> {
        return suspendCancellableCoroutine { continuation ->
            try {
                val request = GetConversationsRequest.newBuilder()
                    .setUserID(userId)
                    .build()
                
                val response = messagingServiceClient?.getConversations(request)
                    ?: throw Exception("Messaging service client not available")
                
                continuation.resume(response.conversationsList)
                
            } catch (error: Exception) {
                continuation.resumeWithException(error)
            }
        }
    }
    
    suspend fun getMessages(conversationId: String, page: Int, pageSize: Int): List<Message> {
        return suspendCancellableCoroutine { continuation ->
            try {
                val request = GetMessagesRequest.newBuilder()
                    .setConversationID(conversationId)
                    .setPage(page.toLong())
                    .setPageSize(pageSize.toLong())
                    .build()
                
                val response = messagingServiceClient?.getMessages(request)
                    ?: throw Exception("Messaging service client not available")
                
                continuation.resume(response.messagesList)
                
            } catch (error: Exception) {
                continuation.resumeWithException(error)
            }
        }
    }
    
    // MARK: - Cleanup
    fun shutdown() {
        try {
            channel?.shutdown()?.awaitTermination(5, TimeUnit.SECONDS)
        } catch (error: Exception) {
            println("Error shutting down gRPC channel: $error")
        }
    }
}

// MARK: - Configuration
data class SonetConfiguration(
    val host: String,
    val port: Int,
    val useTLS: Boolean,
    val timeout: Long
) {
    companion object {
        val development = SonetConfiguration(
            host = "localhost",
            port = 50051,
            useTLS = false,
            timeout = 30L
        )
        
        val staging = SonetConfiguration(
            host = "staging.sonet.app",
            port = 443,
            useTLS = true,
            timeout = 30L
        )
        
        val production = SonetConfiguration(
            host = "api.sonet.app",
            port = 443,
            useTLS = true,
            timeout = 30L
        )
    }
}

// MARK: - Error Types
sealed class SonetError : Exception() {
    object ServiceUnavailable : SonetError() {
        override val message: String = "Service is currently unavailable"
    }
    
    object AuthenticationFailed : SonetError() {
        override val message: String = "Authentication failed"
    }
    
    object NetworkError : SonetError() {
        override val message: String = "Network connection error"
    }
    
    object InvalidResponse : SonetError() {
        override val message: String = "Invalid response from server"
    }
    
    object RateLimited : SonetError() {
        override val message: String = "Too many requests, please try again later"
    }
}